File: src/gui4/windows/kDrive client/kDrive client/ViewModels/App/SyncFileItem.cs
using Infomaniak.kDrive.ServerCommunication.CommStruct;
using Infomaniak.kDrive.Types;
using System;

namespace Infomaniak.kDrive.ViewModels
{

    public class SyncFileItem : UISafeObservableObject
    {
        // SyncFileItem properties
        private NodeType _type;
        private string _path = "";
        private string _newPath = "";
        private string _localNodeId = "";
        private string _remoteNodeId = "";
        private SyncDirection _direction;
        private SyncFileInstruction _instruction;
        private SyncFileStatus _status;
        private ConflictType _conflict;
        private InconsistencyType _inconsistency;
        private CancelType _cancelType;
        private Int64 _size = 0;
        private string _error = "";
        private DateTime _timestamp = DateTime.Now;
        private string _localPath = "";
        private string _parentFolderPath = "";
        private int _progressPercent = 0;
        public Int64 _operationId = 0;

        public Sync Sync { get; }


        public SyncFileItem(Sync sync)
        {
            Sync = sync;
        }

        public SyncFileItem(Sync sync, SyncFileItemInfo info)
        {
            Sync = sync;
            _type = info.Type ?? NodeType.File;
            _path = info.Path ?? string.Empty;
            _localPath = System.IO.Path.Combine(Sync.LocalPath, _path.TrimStart(System.IO.Path.DirectorySeparatorChar));
            _parentFolderPath = System.IO.Path.GetDirectoryName(LocalPath) ?? "";
            _newPath = info.NewPath ?? string.Empty;
            _localNodeId = info.LocalNodeId ?? string.Empty;
            _remoteNodeId = info.RemoteNodeId ?? string.Empty;
            _direction = info.Direction ?? SyncDirection.Unknown;
            _instruction = info.Instruction ?? SyncFileInstruction.None;
            _status = info.Status ?? SyncFileStatus.Unknown;
            _conflict = info.Conflict ?? ConflictType.None;
            _inconsistency = info.Inconsistency ?? InconsistencyType.None;
            _cancelType = info.CancelType ?? CancelType.None;
            _size = info.Size ?? 0;
            _error = info.Error ?? string.Empty;
            _progressPercent = info.Progress ?? 0;
            _operationId = info.OperationId ?? 0;
        }

        public NodeType Type
        {
            get => _type;
            set => SetPropertyInUIThread(ref _type, value);
        }

        public string Path
        {
            get => _path;
            set
            {
                SetPropertyInUIThread(ref _path, value);
                SetPropertyInUIThread(ref _localPath, System.IO.Path.Combine(Sync.LocalPath, _path.TrimStart(System.IO.Path.DirectorySeparatorChar)), nameof(LocalPath));
                SetPropertyInUIThread(ref _parentFolderPath, System.IO.Path.GetDirectoryName(LocalPath) ?? "", nameof(ParentFolderPath));
            }
        }
        public string LocalPath
        {
            get => _localPath;
        }
        public string NewPath
        {
            get => _newPath;
            set => SetPropertyInUIThread(ref _newPath, value);
        }
        public string LocalNodeId
        {
            get => _localNodeId;
            set => SetPropertyInUIThread(ref _localNodeId, value);
        }
        public string RemoteNodeId
        {
            get => _remoteNodeId;
            set => SetPropertyInUIThread(ref _remoteNodeId, value);
        }
        public SyncDirection Direction
        {
            get => _direction;
            set => SetPropertyInUIThread(ref _direction, value);
        }

        public SyncFileInstruction Instruction
        {
            get => _instruction;
            set => SetPropertyInUIThread(ref _instruction, value);
        }

        public bool IsDeletion
        {
            get => Instruction == SyncFileInstruction.Remove;
        }

        public SyncFileStatus Status
        {
            get => _status;
            set => SetPropertyInUIThread(ref _status, value);
        }

        public int ProgressPercent
        {
            get => _progressPercent;
            set => SetPropertyInUIThread(ref _progressPercent, value);
        }

        public ConflictType Conflict
        {
            get => _conflict;
            set => SetPropertyInUIThread(ref _conflict, value);
        }
        public InconsistencyType Inconsistency
        {
            get => _inconsistency;
            set => SetPropertyInUIThread(ref _inconsistency, value);
        }
        public CancelType CancelType
        {
            get => _cancelType;
            set => SetPropertyInUIThread(ref _cancelType, value);
        }
        public string Error
        {
            get => _error;
            set => SetPropertyInUIThread(ref _error, value);
        }

        public Int64 Size
        {
            get => _size;
            set => SetPropertyInUIThread(ref _size, value);
        }

        public DateTime Timestamp
        {
            get => _timestamp;
            set => SetPropertyInUIThread(ref _timestamp, value);
        }

        public Int64 OperationId
        {
            get => _operationId;
            set => SetPropertyInUIThread(ref _operationId, value);
        }

        // Calculated properties
        public string ParentFolderPath
        {
            get => _parentFolderPath;
        }
    }
}

File: src/libcommon/data/syncfileitem.cpp
/*
 * Infomaniak kDrive - Desktop
 * Copyright (C) 2023-2026 Infomaniak Network SA
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "data/syncfileitem.h"

#include "utility/utility.h"

static const auto outParamsType = "type";
static const auto outParamsPath = "path";
static const auto outParamsNewPath = "newPath";
static const auto outParamsLocalNodeId = "localNodeId";
static const auto outParamsRemoteNodeId = "remoteNodeId";
static const auto outParamsDirection = "direction";
static const auto outParamsInstruction = "instruction";
static const auto outParamsStatus = "status";
static const auto outParamsConflict = "conflict";
static const auto outParamsInconsistency = "inconsistency";
static const auto outParamsCancelType = "cancelType";
static const auto outParamsError = "error";
static const auto outParamsSize = "size";
static const auto outParamsProgress = "progress";
static const auto outParamsOperationId = "operationId";

namespace KDC {

SyncFileItem::SyncFileItem(NodeType type, const SyncPath &path, const std::optional<SyncPath> &newPath,
                           const std::optional<NodeId> &localNodeId, const std::optional<NodeId> &remoteNodeId,
                           SyncDirection direction, SyncFileInstruction instruction, SyncFileStatus status, ConflictType conflict,
                           InconsistencyType inconsistency, CancelType cancelType, int64_t size, bool dehydrated) :
    _type(type),
    _path(path),
    _newPath(newPath),
    _localNodeId(localNodeId),
    _remoteNodeId(remoteNodeId),
    _direction(direction),
    _instruction(instruction),
    _status(status),
    _conflict(conflict),
    _inconsistency(inconsistency),
    _cancelType(cancelType),
    _size(size),
    _dehydrated(dehydrated) {}

SyncFileItem::SyncFileItem(NodeType type, const SyncPath &path, const std::optional<NodeId> &localNodeId,
                           const std::optional<NodeId> &remoteNodeId, SyncDirection direction, SyncFileInstruction instruction,
                           ConflictType conflict, int64_t size) :
    _type(type),
    _path(path),
    _localNodeId(localNodeId),
    _remoteNodeId(remoteNodeId),
    _direction(direction),
    _instruction(instruction),
    _conflict(conflict),
    _size(size) {}

void SyncFileItem::toDynamicStruct(Poco::DynamicStruct &dstruct) const {
    CommonUtility::writeValueToStruct(dstruct, outParamsType, _type);
    CommonUtility::writeValueToStruct(dstruct, outParamsPath, Path2QStr(_path).toStdString());
    CommonUtility::writeValueToStruct(dstruct, outParamsNewPath,
                                      _newPath.has_value() ? Path2QStr(_newPath.value()).toStdString() : std::string());
    CommonUtility::writeValueToStruct(dstruct, outParamsLocalNodeId,
                                      _localNodeId.has_value() ? _localNodeId.value() : std::string());
    CommonUtility::writeValueToStruct(dstruct, outParamsRemoteNodeId,
                                      _remoteNodeId.has_value() ? _remoteNodeId.value() : std::string());
    CommonUtility::writeValueToStruct(dstruct, outParamsDirection, _direction);
    CommonUtility::writeValueToStruct(dstruct, outParamsInstruction, _instruction);
    CommonUtility::writeValueToStruct(dstruct, outParamsStatus, _status);
    CommonUtility::writeValueToStruct(dstruct, outParamsConflict, _conflict);
    CommonUtility::writeValueToStruct(dstruct, outParamsInconsistency, _inconsistency);
    CommonUtility::writeValueToStruct(dstruct, outParamsCancelType, _cancelType);
    CommonUtility::writeValueToStruct(dstruct, outParamsError, _error);
    CommonUtility::writeValueToStruct(dstruct, outParamsSize, _size);
    CommonUtility::writeValueToStruct(dstruct, outParamsProgress, _progress);
    CommonUtility::writeValueToStruct(dstruct, outParamsOperationId, _operationId);
}

} // namespace KDC
